secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/sign/SignPayloadUtil.java

```diff
@@ -0,0 +1,104 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.sapi.gateway.sign;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.forgerock.json.jose.builders.JwtBuilderFactory;
+import org.forgerock.json.jose.builders.SignedJwtBuilderImpl;
+import org.forgerock.json.jose.jws.JwsAlgorithm;
+import org.forgerock.json.jose.jws.SignedJwt;
+import org.forgerock.json.jose.jws.SigningManager;
+import org.forgerock.json.jose.jws.handlers.SigningHandler;
+import org.forgerock.json.jose.jwt.JwtClaimsSet;
+import org.forgerock.secrets.NoSuchSecretException;
+import org.forgerock.secrets.Purpose;
+import org.forgerock.secrets.SecretsProvider;
+import org.forgerock.secrets.keys.SigningKey;
+import org.forgerock.util.Reject;
+import org.forgerock.util.promise.Promise;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Util to sign objects <br/>
+ * Implementation of {@link SignUtil}
+ */
+public class SignPayloadUtil implements SignUtil {
+
+    private static final Logger logger = LoggerFactory.getLogger(SignPayloadUtil.class);
+    private final SigningManager signingManager;
+    private final Purpose<SigningKey> signingKeyPurpose;
+    private final Map<String, Object> critHeaderClaims;
+
+    private final String signingKeyId;
+    private final String algorithm;
+    private final String kid;
+    public SignPayloadUtil(
+            SecretsProvider secretsProvider,
+            Map<String, Object> critHeaderClaims,
+            String signingKeyId,
+            String kid,
+            String algorithm
+    ) {
+        Reject.ifNull(secretsProvider, "secretsProvider must be supplied");
+        Reject.ifNull(kid, "kid must be supplied");
+        Reject.ifNull(signingKeyId, "signingKeyId must be supplied");
+        Reject.ifNull(algorithm, "algorithm must be supplied");
+        this.signingManager = new SigningManager(secretsProvider);
+        this.signingKeyId = signingKeyId;
+        this.kid = kid;
+        this.critHeaderClaims = critHeaderClaims == null ? Collections.EMPTY_MAP : critHeaderClaims;
+        this.algorithm = algorithm;
+        this.signingKeyPurpose = Purpose.purpose(signingKeyId, SigningKey.class);
+    }
+
+
+    @Override
+    public String sign(final Map<String, Object> payload) {
+        Reject.ifNull(payload, "payload must be supplied");
+        Reject.ifTrue(payload.isEmpty(),"payload map must not be empty");
+        logger.debug("Payload to be signed:\n {}\n", payload);
+        Promise<SigningHandler, NoSuchSecretException> signingHandler = signingManager.newSigningHandler(signingKeyPurpose);
+        try {
+            final JwtClaimsSet jwtClaimsSet = new JwtClaimsSet(payload);
+
+            SignedJwtBuilderImpl signedJwtBuilder = new JwtBuilderFactory()
+                    .jws(signingHandler.getOrThrow())
```

Comment:
Calls to any of the Promise.get* methods must be avoided in production code. 

All of the get calls are blocking which is an issue in IG as it uses the reactor pattern to service multiple concurrent requests off a single (or small number of threads). Any blocking calls will therefore prevent the thread from making progress and can even lead to deadlock.

Whenever are Promise is used within a method then typically that method needs to return a Promise itself, please rewrite this method so that it returns a Promise<String, SomeException>. This will in turn require that the calling code handles the Promise in an async fashion

Reply:
Implementation fixed to handle properly the Promise result and return a simple string or throw and exception if happens.